### PR TITLE
feat: add compiler v2 explanations

### PR DESCRIPTION
**Pull request by enitrat**

### CODE REVIEW

File: src/ch99-01-03-00-a-deeper-dive-into-contracts.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# A deeper dive into contracts
+
+In the previous section, we gave an introductory example of a smart contract written in Cairo. In this section, we'll be taking a deeper look at all the components of a smart contract, step by step.
+
+In the previous chapter, we talked about _interfaces_, specified the difference between _public functions, external functions and view functions_, and we mentioned how to interact with _storage_.
```
Comment (rodrigo-pino) [inline src/ch99-01-03-00-a-deeper-dive-into-contracts.md:5]: Both paragraphs start the same way. Perhaps the second one could start:
"When we discussed \[interfaces\]\(link\) ..."

File: src/ch99-01-03-00-a-deeper-dive-into-contracts.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+At this point, you should have multiple questions that come to mind:
+
+- How do I define internal/private functions?
+- How can I emit events? How can I index them?
+- Where should I defined functions that do not need to access the contract's state?
+- Is there a way to reduce the boilerplate?
+- How can I store more complex data types?
+
+Luckily, we'll be answering all these questions in this chapter. To answer these questions, let's consider the following example contract that we'll be using throughout this chapter:
+
```
Comment (rodrigo-pino) [inline src/ch99-01-03-00-a-deeper-dive-into-contracts.md:16]: > To answer these questions,

I would remove that part of the sentence

File: src/ch99-01-03-03-contract-events.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+The `#[derive(starknet::Event)]` attribute causes the compiler to generate an implementation for the above trait,
+instantiated with the Event type, which in our example is the following enum:
+
+```rust
+{{#include ../listings/ch99-starknet-smart-contracts/listing_99_03_example_contract.cairo:event}}
+```
+
+Each event variant has to be a struct of the same name as the variant, and each variant needs to implement the `starknet::Event` trait itself.
+Moreover, the members of these variants must implement the `Serde` trait, as keys/data are added to the event using a serialization process.
```
Comment (rodrigo-pino) [inline src/ch99-01-03-03-contract-events.md:25]: I would add a link to Appendix C: Serializing with Serde

**Review by rodrigo-pino — COMMENTED** [review]
Very thorough!

File: src/ch99-01-03-00-a-deeper-dive-into-contracts.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
+# A deeper dive into contracts
+
+In the previous section, we gave an introductory example of a smart contract written in Cairo. In this section, we'll be taking a deeper look at all the components of a smart contract, step by step.
+
+In the previous chapter, we talked about _interfaces_, specified the difference between _public functions, external functions and view functions_, and we mentioned how to interact with _storage_.
+
+At this point, you should have multiple questions that come to mind:
+
+- How do I define internal/private functions?
+- How can I emit events? How can I index them?
```
Comment (fricoben) [inline src/ch99-01-03-00-a-deeper-dive-into-contracts.md:10]: indexing is not really part of the contract but could be cool to have some words on it as it's something really important

File: src/ch99-01-03-01-storage-variables.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
+## Storage Variables
+
+As stated previously, storage variables allow you to store data that will be stored on the blockchain in the contract's storage. These data are persistent and can be accessed and modified anytime once the contract is deployed.
```
Comment (fricoben) [inline src/ch99-01-03-01-storage-variables.md:3]: `on the blockchain in the contract's storage` ==> ` in the contract's storage that is itself stored on the blockchain`

top-down better imo

File: src/ch99-01-03-01-storage-variables.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+<span class="caption">Listing 99-2: A Storage Struct</span>
+
+The storage struct is a [struct](./ch04-00-using-structs-to-structure-related-data.md) like any other,
+except that it **must** be annotated with `#[storage]` allowing you to store mappings using the `LegacyMap` type.
+
+### Storing Mappings
+
+Mappings are a key-value data structure that you can use to store data within a smart contract. They are essentially hash tables that allow you to associate a unique key with a corresponding value. Mappings are also useful to store sets of data, as it's impossible to store arrays in storage.
+
+A mapping is a variable of type `LegacyMap`, in which the key and value types are specified within angular brackets <>.
```
Comment (fricoben) [inline src/ch99-01-03-01-storage-variables.md:20]: <> ==> `<>`

File: src/ch99-01-03-01-storage-variables.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+It is important to note that the `LegacyMap` type can only be used inside the `Storage` struct, and can't be used to define mappings in user-defined structs.
+
+You can also create more complex mappings than that found in Listing 99-2 like the popular `allowances` storage variable in the ERC20 Standard which maps the `owner` and `spender` to the `allowance` using tuples:
```
Comment (fricoben) [inline src/ch99-01-03-01-storage-variables.md:23]: `than that found` ==> `than that, you can find one`

File: src/ch99-01-03-03-contract-events.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
+## Events
+
+Events are custom data structures that are emitted by smart contracts during execution.
+They provide a way for smart contracts to communicate with the external world by logging information
+about specific occurrences in a contract.
+
+Events play a crucial role in the creation of smart contracts. Take, for instance, the Non-Fungible Tokens (NFTs) minted on Starknet. All of these are indexed and stored in a database, then displayed to users through the use of these events. Neglecting to include an event within your NFT contract could lead to a bad user experience. This is because users may not see their NFTs appear in their wallets (wallets use these indexers to display a user's NFTs).
```
Comment (fricoben) [inline src/ch99-01-03-03-contract-events.md:7]: Gud example

File: src/ch99-01-01-introduction-to-smart-contracts.md [hunk 9]
```diff
@@ -43,6 +43,25 @@ As Ethereum continues to mature, we can expect the use cases and applications of
...
+A famous theorem in the blockchain space states that it is impossible to achieve a high level of scalability, decentralization, and security simultaneously; trade-offs must be made. Ethereum is at the intersection of decentralization and security. Eventually, it was decided that Ethereum's purpose would be to serve as a secure settlement layer, while complex computations would be offloaded to other networks built on top of Ethereum. These are called Layer 2s (L2s).
+
+The two primary types of L2s are optimistic rollups and validity rollups. Both approaches involve compressing and batching numerous transactions together, computing the new state, and settling the result on Ethereum (L1). The difference lies in the way the result is settled on L1. For optimistic rollups, the new state is considered valid by default, but there is a 7-day window for nodes to identify malicious transactions.
```
[the remark this answers is not in the document]
Comment (enitrat) [reply src/ch99-01-01-introduction-to-smart-contracts.md:50]: Adding link to the starknet architecture docs

File: src/ch99-01-01-introduction-to-smart-contracts.md [hunk 10]
```diff
@@ -43,6 +43,25 @@ As Ethereum continues to mature, we can expect the use cases and applications of
...
+
+Remember Cairo? It is, in fact, a language developed specifically to work with STARKs and make them general-purpose. With Cairo, we can write **provable code**. In the context of Starknet, this allows proving the correctness of computations from one state to another.
+
+Unlike most (if not all) of Starknet's competitors that chose to use the EVM (either as-is or adapted) as a base layer, Starknet employs its own VM. This frees developers from the constraints of the EVM, opening up a broader range of possibilities. Coupled with decreased transaction costs, the combination of Starknet and Cairo creates an exciting playground for developers. Native account abstraction enables more complex logic for accounts and transaction flows. Emerging use cases include **transparent AI** and artificial intelligence and machine learning applications. Finally, **blockchain games** can be developed entirely **on-chain**. Starknet has been specifically designed to maximize the capabilities of STARK proofs for optimal scalability.
```
[the remark this answers is not in the document]
Comment (enitrat) [reply src/ch99-01-01-introduction-to-smart-contracts.md:56]: Added link to Starknet Book as well

File: src/ch99-01-03-00-a-deeper-dive-into-contracts.md (hunk 4 above)
Comment (enitrat) [reply src/ch99-01-03-00-a-deeper-dive-into-contracts.md:10]: I meant, index them by their keys.

File: listings/ch99-starknet-smart-contracts/listing_99_01.cairo [hunk 11]
```diff
@@ -0,0 +1,36 @@
+// ANCHOR:all
```
[the remark this answers is not in the document]
Comment (enitrat) [reply listings/ch99-starknet-smart-contracts/listing_99_01.cairo:1]: This is used to only render specific parts of files in the rendering of the markdown pages. It's not very important as the reader doesn't directly see this

File: listings/ch99-starknet-smart-contracts/listing_99_01.cairo [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    use starknet::get_caller_address;
+    use starknet::ContractAddress;
+
+    #[storage]
+    struct Storage {
+        stored_data: u128
+    }
+
+    //ANCHOR: impl
+    #[external(v0)]
```
[the remark this answers is not in the document]
Comment (enitrat) [reply listings/ch99-starknet-smart-contracts/listing_99_01.cairo:21]: It's in the second page

**Pull request merged**